interfaces/builtin/shared_memory.go

```diff
@@ -28,36 +28,74 @@ import (
 	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/snap"
 )
 
 const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
 
-// The plug side of shared-memory implements auto-connect to a matching slot -
-// this is permitted even though the interface is super-privileged because using
-// a slot requires a store declaration anyways so just declaring a plug will not
-// grant access unless a slot was also granted at some point.
+// The plug side of shared-memory can operate in two modes: if the
+// private attribute is set to true, then it can be connected to the
+// implicit system slot to be given a private version of /dev/shm.
+//
+// For a plug without that attribute set, it will connect to a
+// matching application snap slot - this is permitted even though the
+// interface is super-privileged because using a slot requires a store
+// declaration anyways so just declaring a plug will not grant access
+// unless a slot was also granted at some point.
 const sharedMemoryBaseDeclarationPlugs = `
   shared-memory:
     allow-installation: true
     allow-connection:
-      slot-attributes:
-        shared-memory: $PLUG(shared-memory)
+      -
+        plug-attributes:
+          private: false
+        slot-attributes:
+          shared-memory: $PLUG(shared-memory)
+      -
+        plug-attributes:
+          private: true
+        slot-snap-type:
+          - core
     allow-auto-connection:
-      slot-publisher-id:
-        - $PLUG_PUBLISHER_ID
-      slot-attributes:
-        shared-memory: $PLUG(shared-memory)
+      -
+        plug-attributes:
+          private: false
+        slot-publisher-id:
+          - $PLUG_PUBLISHER_ID
+        slot-attributes:
+          shared-memory: $PLUG(shared-memory)
+      -
+        plug-attributes:
+          private: true
+        slot-snap-type:
+          - core
 `
 
-// shared-memory slots are super-privileged and thus denied to any snap except
-// those that get a store declaration to do so, but the intent is for
-// application or gadget snaps to use the slot much like the content interface.
+// shared-memory slots can appear either as an implicit system slot,
+// or as a slot on an application snap.
+//
+// The implicit version of the slot is intended to auto-connect with
+// plugs that have the private attribute set to true.
+//
+// Slots on app snaps connect to non-private plugs. They are are
+// super-privileged and thus denied to any snap except those that get
+// a store declaration to do so, but the intent is for application or
+// gadget snaps to use the slot much like the content interface.
 const sharedMemoryBaseDeclarationSlots = `
   shared-memory:
-    allow-installation: false
-    deny-connection: true
-    deny-auto-connection: true
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
```

Comment:
the easy way to implement what is asked for where slots on app snaps are denied but the implicit slot is okay would be to change this to:

```yaml
  shared-memory:
    allow-installation:
      slot-snap-type:
        - core
    deny-connection: false
```

then app snap slots evaluate to false for allow-installation as they are not a snap of type `core`

Reply:
I had tried this version, and found that it caused the `interfaces-shared-memory` spread test to fail.

It seems that a locally installed snap with a shared-memory slot can be installed when the base declaration says `allow-installation: false`, but not with `allow-installation: {slot-snap-type: [core]}`. That would essentially make those snaps impossible to test before publishing them to the store.

Reply:
That sounds like it should do the trick.

As I understand it, the slot base declaration would then read something like:

```
  shared-memory:
    allow-installation:
      slot-snap-id:
        - 99T7MUlRhtI3U0QFgl5mXXESAiSwt776
        - PMrrV4ml8uWuEUDBT8dSGnKUYbevVhc4
 ```

(i.e. snap IDs of `core` and `snapd`). An unasserted app snap should still be able to instal with a shm slot, and the store can override this for a snap with a snap-declaration assertion.